### DISCUSSION TITLE
n_chains parameter in the sampler

### DISCUSSION THREAD

**Discussion by abukva** [netket/netket#694] [Q&A] [answered]
Hi, first thanks for this great project it makes a life of physicist so much easier :D I was following along  a simple 1D ising in the example folder and there is a n_chains parameters in a sampler that a documentation says "Number of batches along the chain" I don't quite understand what this means. I've tried changing it and it does have an impact on performance but I am not able to understand how exactly. I am working with a GPU version of netkat. Thanks!

**Comment by gcarleo** [comment, marked as answer]
Hi, maybe [this ](https://www.netket.org/docs/varstate.html#sampling)can be helpful?

**Reply by abukva** [reply]
Aha so that specifies how many parallel samplers are run to speed up sampling, but I need to be careful to have enough samples in  one chain to have a meaningful statistics? And when I run on GPU let's say cuda core vise does this mean that only n_chains number of cores will actually do work?

**Reply by gcarleo** [reply]
Yes those are independent samplers running in parallel on the same device (in your case on the GPU). And yes you should be careful to have enough samples per independent chain. On the other hand no, there is no restrictions on the number of GPU processes that are run, in principle these are low level details that are set by JAX. In any case, in practice on the GPU it is much more efficient to use a large number of n_chains... (say at least n_chains=64, but that strongly depends on your model/system size)

**Reply by abukva** [reply]
Thanks a lot!

**Reply by abukva** [reply]
@gcarleo just out of curiosity where I could read more about how JAX does this distribution over cores in order to get the maximum performance form my simulation? Thanks again for the help!

**Reply by PhilipVinc** [reply]
If you want to read about how Jax uses different CPU cores, there is very little written. If you search issues in the jax repository with the keywords `cpu multithreading` or `MPI` or `taskset`  you might find something but it's really not well documented.
About usage of GPU cores, that is very hardcore things you are looking at and you won't find anything easy to digest. You can investigate it by yourself, however, by profiling the jax program and checking the GPU usage with `nvprof` or tensor board (see jax documentation). 

About how to properly set the netket sampler settings to achieve good performance, I'm writing some notes that will end up in the documentation eventually.

**Reply by abukva** [reply]
Thank you for the info, looking forward to those notes!
